File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        
        
        int ans=0;
        
        int leftMax=0,rightMax=0;
        
        int i=0,j=height.length-1;
        
        while(i<=j)
        {
            if(leftMax<rightMax)
            {
                leftMax=Math.max(leftMax,height[i]);
                ans+=leftMax-height[i];
                i++;
            }
            else
            {
                rightMax=Math.max(rightMax,height[j]);
                ans+=rightMax-height[j];
                j--;
            }
        }
        return ans;
        
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        
        int ans=0;
        
        int l=0,r=height.size()-1;
        int leftMax=height[0];
        int rightMax=height[height.size()-1];
        
        while(l<r)
        {
            if(leftMax<rightMax)
            {
                ans+=max(leftMax-height[l+1],0);
                l++;
                leftMax=max(leftMax,height[l]);
            }
            else
            {
                ans+=max(rightMax-height[r-1],0);
                r--;
                rightMax=max(rightMax,height[r]);
            }
        }
        return ans;
        
    }
};